fashion-hyperparameter/optimize_class.py

```diff
@@ -0,0 +1,106 @@
+import os
+from collections import defaultdict
+from itertools import tee
+import json
+
+from jina.flow import Flow
+from jina.helper import colored
+from jina.logging import default_logger as logger
+
+
+class Optimizer:
+    def __init__(self, index_yaml, evaluate_yaml, 
+                       index_document_generator, evaluation_document_generator,
+                       index_batch_size, query_batch_size,
+                       trial_parameter_sampler,
+                       n_trials,
+                       direction='maximize', seed=42,
+                       config_dir='config', best_config_filename='best_config.json'):
+        self.index_yaml = index_yaml
+        self.evaluate_yaml = evaluate_yaml
+        self.index_document_generator = index_document_generator
+        self.evaluation_document_generator = evaluation_document_generator
+        self.index_batch_size = index_batch_size
+        self.query_batch_size = query_batch_size
+        self.trial_parameter_sampler = trial_parameter_sampler
+        self.n_trials = n_trials
+        self.direction = direction
+        self.seed = seed
+        self.config_dir = config_dir
+        self.best_config_filename = best_config_filename
+
+
+    def run_indexing(self):
+        if os.path.exists(os.environ['JINA_WORKSPACE']):
+            print(colored('--------------------------------------------------------', 'red'))
+            print(colored('----- Workspace already exists. Skipping indexing. -----', 'cyan'))
+            print(colored('--------------------------------------------------------', 'red'))
+            return
+
+        self.index_document_generator, index_document_generator_ = tee(self.index_document_generator)
+
+        with Flow().load_config(self.index_yaml) as f:
+            f.index(index_document_generator_, batch_size=self.index_batch_size)
+
+    def run_querying(self, callback):
+        self.evaluation_document_generator, evaluation_document_generator_ = tee(self.evaluation_document_generator)
+
+        with Flow().load_config(self.evaluate_yaml) as f:
+            f.search(
+                evaluation_document_generator_,
+                batch_size=self.query_batch_size,
+                output_fn=callback
+            )
+
+    @staticmethod
+    def parameters_to_env(parameters):
+        for environment_variable, value in parameters.items():
+            os.environ[environment_variable] = str(value)
+
+    def run_evaluation(self, parameters, evaluation_callback):
+        Optimizer.parameters_to_env(parameters)
+        self.run_indexing()
+        self.run_querying(evaluation_callback)
```

Comment:
so for instance, you could do `self.run(pre_hook=self.index(), post_hook=evaluation_callback())` or something in these lines

Reply:
Does there exist any situation where pre_hook is different than index?

Reply:
Does there exist any situation where post_hook is different than eval?

Reply:
It could be empty. 

I am not meaning to put it in the example. But when designing the module classes to have it in mind. Another thing is how u expose these features to the user.

Reply:
for post_hook different than eval it may make sense to run evaluation by default yes. 

I am just raising awareness for the definitive design to keep that in mind. We should be coherent with the idea that `evaluation` is agnostic to the `flow` we are using

Reply:
Perhaps we should just give in a list of flows with there respective evaluation callbacks. Then we guarantee full flexibility.

Reply:
Yes, but would be important to have different 'pre-processing' (here run an index flow) per each evaluation? Or this preprocessing could be shared by all the Flows in the Pool?